### PR TITLE
Initialization

### LINKED ISSUE #3: Change the parameters ?
function kmeans(data, k, options)

options.initialisation
* array of your current centers
* 'random' (default value) : we take randomly k points of the data

### PR DESCRIPTION
**Pull request by maasencioh**
There are 3 possible ways to initialice currently the points:
  * K initial points in the format [x,y,z,...]
  * `'random'` will choose K random different values
  * `'moreDistant'` will choose the more distant points to a first random pick

closes #3 

cc @mljs/collaborators

### CODE REVIEW

File: src/initialization.js [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+const distance = require('ml-distance-euclidean');
+
+/**
+ * Choose K different random points from the original data
+ * @ignore
+ * @param {Array<Array<Number>>} data - Points in the format to cluster [x,y,z,...]
+ * @param {Number} K - Number of clusters
+ * @return {Array<Array<Number>>} - Initial random points
+ */
+function random(data, K) {
```
Comment (targos) [inline src/initialization.js]: This probably belongs outside of this project. Or maybe there is already something on npm for that ?

File: src/initialization.js [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    let len = data.length;
+    let ans = new Array(K);
+
+    for (let i = 0; i < K; ++i) {
+        let rand = Math.floor(Math.random() * len);
```
Comment (targos) [inline src/initialization.js]: You use this expression 3 times in the code. Could be factored out in a function

File: src/initialization.js [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+/**
+ * Chooses the more distant points to a first random pick
+ * @ignore
+ * @param {Array<Array<Number>>} data - Points in the format to cluster [x,y,z,...]
+ * @param {Number} K - Number of clusters
+ * @return {Array<Array<Number>>} - Initial random points
+ */
+function moreDistant(data, K) {
```
Comment (targos) [inline src/initialization.js]: mostDistant

File: src/initialization.js [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+ * @return {Array<Array<Number>>} - Initial random points
+ */
+function random(data, K) {
+    let len = data.length;
+    let ans = new Array(K);
+
+    for (let i = 0; i < K; ++i) {
+        let rand = Math.floor(Math.random() * len);
+
+        /* istanbul ignore next */
```
Comment (targos) [inline src/initialization.js]: Please don't use `istanbul ignore` in this file.

File: src/kmeans.js [hunk 5]
```diff
@@ -1,148 +1,80 @@
...
-}
+const utils = require('./utils');
+const init = require('./initialization');
 
 const defaultOptions = {
     maxIterations: 100,
     tolerance: 1e-6,
-    withIterations: false
+    withIterations: false,
+    initialization: 'moreDistant'
```
Comment (targos) [inline src/kmeans.js]: mostDistant

File: src/kmeans.js [hunk 6]
```diff
@@ -1,148 +1,80 @@
...
+    } else {
+        switch (options.initialization) {
+            case 'random':
+                centers = init.random(data, K);
+                break;
+            case 'moreDistant':
+                centers = init.moreDistant(data, K);
+                break;
+            default:
+                throw new Error('Unknown initialization method');
```
Comment (targos) [inline src/kmeans.js]: Can you put the value of `options.initialization` in the message ?

File: src/utils.js [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+ * @param {Array<Array<Number>>} data - the [x,y,z,...] points to cluster
+ * @param {Array<Array<Number>>} centers - the K centers in format [x,y,z,...]
+ * @param {Array<Number>} clusterID - the cluster identifier for each data dot
+ * @returns {Number} the sum of squared errors
+ */
+function computeSSE(data, centers, clusterID) {
+    let sse = 0;
+    let nData = data.length;
+    let c = 0;
+    for (let i = 0; i < nData; i++) {
```
Comment (targos) [inline src/utils.js]: Mutation of `let` bindings in loops is still very slow in V8. Can you please only use `var` or `const` in hot code?

File: src/utils.js [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+ * @param {Array<Array<Number>>} data - the [x,y,z,...] points to cluster
+ * @param {Array<Array<Number>>} centers - the K centers in format [x,y,z,...]
+ * @returns {Array} the cluster identifier for each data dot
+ */
+function updateClusterID(data, centers) {
+    let nData = data.length;
+    let k = centers.length;
+    let aux = 0;
+    let clusterID = new Array(nData);
+    for (let i = 0; i < nData; i++)
```
Comment (targos) [inline src/utils.js]: Please use braces or put everything on one line.

File: src/utils.js [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+        d[i] = new Array(k);
+        for (let j = 0; j < k; j++) {
+            aux = squaredDistance(data[i], centers[j]);
+            d[i][j] = new Array(2);
+            d[i][j][0] = aux;
+            d[i][j][1] = j;
+        }
+        let min = d[i][0][0];
+        let id = 0;
+        for (let j = 0; j < k; j++)
```
Comment (targos) [inline src/utils.js]: Please use braces. Same remark for the rest of the file

File: src/kmeans.js [hunk 10]
```diff
@@ -1,148 +1,80 @@
...
-    let withIterations = options.withIterations || defaultOptions.withIterations;
+function kmeans(data, K, options) {
+    options = Object.assign({}, defaultOptions, options);
 
     let nData = data.length;
-    if (nData === 0) {
-        return [];
+
+    if (K > nData) {
+        throw new Error('The numbers in data should be bigger than the k value');
```
Comment (targos) [inline src/kmeans.js]: I'd write "Data length should be bigger than K"

File: src/kmeans.js [hunk 11]
```diff
@@ -1,148 +1,80 @@
...
+    }
+
+    let centers;
+    if (Array.isArray(options.initialization)) {
+        if (options.initialization.length !== K) {
+            throw new Error('The initial centers should have the same length than K');
```
Comment (targos) [inline src/kmeans.js]: same length as K

File: src/utils.js [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+    let aux = 0;
+    let clusterID = new Array(nData);
+    for (let i = 0; i < nData; i++)
+        clusterID[i] = 0;
+    let d = new Array(nData);
+    for (let i = 0; i < nData; i++) {
+        d[i] = new Array(k);
+        for (let j = 0; j < k; j++) {
+            aux = squaredDistance(data[i], centers[j]);
+            d[i][j] = new Array(2);
```
Comment (targos) [inline src/utils.js]: d[i][j] = [aux, j];

**Comment by maasencioh** [comment]
@targos all the changes were made as discussed

**Pull request merged**